crates/ruff/src/rules/ruff/rules/unreachable.rs

```diff
@@ -0,0 +1,964 @@
+            self.bits[n] |= 1 << s;
+            false
+        } else {
+            true
+        }
+    }
+
+    /// Returns an iterator of all unset indices.
+    fn unset(&self) -> impl Iterator<Item = usize> + '_ {
+        let mut index = 0;
+        let mut shift = 0;
+        let last_max_shift = self.capacity % usize::BITS as usize;
+        iter::from_fn(move || loop {
+            if shift >= usize::BITS as usize {
+                shift = 0;
+                index += 1;
+            }
+            if self.bits.len() <= index || (index >= self.bits.len() - 1 && shift >= last_max_shift)
+            {
+                return None;
+            }
+
+            let is_set = (self.bits[index] & (1 << shift)) != 0;
+            shift += 1;
+            if !is_set {
+                return Some((index * usize::BITS as usize) + shift - 1);
+            }
+        })
+    }
+}
+
+/// Set bits in `reached_map` for all blocks that are reached in `blocks`
+/// starting with block at index `idx`.
+fn mark_reached(reached_map: &mut Bitmap, blocks: &[BasicBlock<'_>], mut idx: BlockIndex) {
+    loop {
+        let block = &blocks[idx];
+        if reached_map.set(idx) {
+            return; // Block already visited, no needed to do it again.
+        }
+
+        match &block.next {
+            NextBlock::Always(next) => idx = *next,
+            NextBlock::If {
+                condition,
+                next,
+                orelse,
+            } => {
+                match taken(condition) {
+                    Some(true) => idx = *next,    // Always taken.
+                    Some(false) => idx = *orelse, // Never taken.
+                    None => {
+                        // Don't know, both branches might be taken.
+                        idx = *next;
+                        mark_reached(reached_map, blocks, *orelse);
+                    }
+                }
+            }
+            NextBlock::Terminate => return,
+        }
+    }
+}
+
+/// Determines if `condition` is taken.
+/// Returns `Some(true)` if the condition is always true, e.g. `if True`, same
+/// with `Some(false)` if it's never taken. If it can't be determined it returns
+/// `None`, e.g. `If i == 100`.
+fn taken(condition: &Condition) -> Option<bool> {
+    // TODO: add more cases to this where we can determine a condition
+    // statically. For now we only consider constant booleans.
+    match condition {
+        Condition::Test(expr) => match expr {
+            Expr::Constant(constant) => constant.value.as_bool().copied(),
+            _ => None,
+        },
+        Condition::Iterator(_) => None,
+        Condition::Match { .. } => None,
+    }
+}
+
+/// Collection of basic block.
+#[derive(Debug, PartialEq)]
+struct BasicBlocks<'stmt> {
+    /// # Notes
+    ///
+    /// The order of these block is unspecified. However it's guaranteed that
+    /// the last block is the statement in the function and the first block is
```

Comment:
I don't understand this part. Which statement is *the statement*? And how's the *last statement* in a function defined?

Reply:
> I don't understand this part. Which statement is _the statement_?

While changing the docs I forgot a crucial word:

```suggestion
    /// the last block is the first statement in the function and the first block is
```

> And how's the _last statement_ in a function defined?

It's whatever the last statements in in the list of `Stmt` in the call to `BasicBlocks::from(&stmts)`. In other words the last line of code in the function (at the 1st level of indentation, i.e. the function body level of indentation). For example

```python
def function():
    while True: # Last line.
        continue # (part of the `while` statement)
```

Reply:
Oh interesting that the blocks are in reverse order. What's the motivation for adding them in reverse order?

Reply:
I saw that you answered this in your reply further down.